README.md
# Galactic Guardians:
# The Sentinel Saga
## _== Space arcade ==>>_
![Galactic Guardians: The Sentinel Saga](https://github.com/Rexant-b2k/space_arcade/blob/main/screenshots/splash.webp)
## based on:
[![N|Pygame](https://www.pygame.org/docs/_static/pygame_tiny.png)](https://www.pygame.org/)

This game, inspired by classic 80's arcades, will deep you into old-fashion fun and joy.
![Game process](https://github.com/Rexant-b2k/space_arcade/blob/main/screenshots/screen2.webp)
![Game pause](https://github.com/Rexant-b2k/space_arcade/blob/main/screenshots/screen1.webp)
![Start screen](https://github.com/Rexant-b2k/space_arcade/blob/main/screenshots/screen3.webp)

## Features
Project developing currently in progress. However the current version is fully playable.

- Three types of enemy ships
- Lasershots and fun
- Protect your base!

## Controls
| Button           | Action       |
| ---------------- | ------------ |
| A or Left Arrow  | Move left    |
| D or Right Arrow | Move Right   |
| W or Up Arrow    | Move Forward |
| S or Down Arrow  | Move Back    |
| Spacebar         | Shoot        |
| P                | Pause        |


## Installation and start the game
*Requires python and pygame library. I recommend use virtual enviroment (venv)

1. Clone this repository
2. Perform commands to prepare venv and download dependencies
```sh
python3 -m venv venv
python -m pip install --upgrade pip
source venv/bin/activate
pip install -r requirements.txt
```
3. Start the game
```sh
python main.py
```
4. Enjoy!

## Tech
Galactic Guardians uses a number of open source and proprietary projects to work properly:
- [Python] - Python 3.11
- [Pygame] - Pygame library  
  
And of course Galactic Guardians: The Sentinel Saga itself is open source with a [public repository][Space_arcade]
 on GitHub.

## Plugins

Galactic Guardians: The Sentinel Saga is currently extended with the following plugins.
Instructions on how to use them in your own application are linked below.

| Plugin       | README          |
| -------------| --------------- |
| Dillinger    | [Dillinger.io]  |

## Version history
v0.3a (current) - Added medkits which restore you 50% of your health. Tech: Functional improvements, new module "helper" and reassembling game code. [12.10.2023]  
v0.2a - Lasers are not belongs ships anymore. Now if you defeat an enemy - his laser will still go on (previously - dissapered with the ship). Tech: Classes inheratances and global change to OOP approach. [16.09.2023]  
v0.1a - Base gameplay, initial refactoring of code [08.09.2023]  

## Authors
[Sergei Baryshevskii](https://www.linkedin.com/in/barysecho/)  
With gratitude to:  
[Tim Ruscica](https://www.techwithtim.net/) (approach to game development powered by pygame)

## License

BSD-3 Clause License

**Free Software, Hello everybody**

[//]: # (These are reference links used in the body of this note and get stripped out when the markdown processor does its job. There is no need to format nicely because it shouldn't be seen. Thanks SO - http://stackoverflow.com/questions/4823468/store-comments-in-markdown-syntax)

   [Space_arcade]: <https://github.com/Rexant-b2k/space_arcade/>
   [git-repo-url]: <https://github.com/Rexant-b2k/space_arcade.git>
   [Pygame]: <https://www.pygame.org/>
   [Python]: <https://www.python.org/>
   [Dillinger.io]: <https://dillinger.io/>




CHANGELOG.MD

# Change Log
## Galactic Guardians: The Sentinel Saga
All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).
 
## [0.3-alpha] - 2023-12-12
 
Here I am writing upgrading notes for brands. It's my effort to make them as
straightforward as possible.
 
### Added
- 'Game init' function and entity with same name, which keeps base game settings to launch the game.
- 'get screen res' function, based on 3rd party screeninfo module, which initialize all displays and find their resolutions and the primary one.
- 'Escape game' function to have an ability to leave the game.
- Helper module to move there callable functions from main module.
- Medkits! Now you can restore your health/hull. If damaged, the probability of appearing medkit is increased.

 
### Changed
  
#### Tech
    
- 'Game init' entity stores seetings of of resolution, screen refresh constant, initialized game window, and paths to background and menubackground splash screens with applied transformations to fit current resolution.
- 'get screen resolution' collects info about all connected displays, their resolutions and which of them is primary. If no primary (in a case of Virtual Machine) - uses the first one available. Basing on screen resolution, finds the most proper game resolution (maximum closest (equal or smaller one) to current display resolution). Analyzes if display is squared, widescreen or superwide and calculate best game window size from _minimal_ parameter.
- Since all objects in game are 'SpaceObjects' inheritors (since 0.2a), and collision is possible only between this type of objects, function 'collide' was removed from global scope of data module and moved to SpaceObject as a method. All nessesary changes in main and helper modules were made.
- Replaced 'paused' and 'escape game' reaction to _pygame.Keydown_ from _pygame.key.get_pressed()_
- Enemy generation is now made as a dedicated function and moved to helper module
- Enemy movements are now made as a dedicated function and moved to helper module
- Weapon shell movements are now made as a dedicated function and moved to helper module
- Changed the order in WS_RESOLUTION constant to help 'get_screen_res' function.
- Added helpers to find the middle horizontal, vertial and total position on screen.
- screeninfo module adeed to requirements.txt

#### Gameplay

- 'Escape game' (ESC button) can now be pressed in main menu
- Player movement speed changed to variable which depends on screen resolution. Now, it is a constant time to move player from left edge position to the right one, regardless of resolution.
- background pictures are now being transformed to correspond current resolution
- Enemy ships don't attack in bottom part of screen (30%) anymore to reduce instant death of player by unexpected shots.

### Fixed
- 'Pause' function no more switch on/off rapidly following game pressing events. ('Pygame' usually copies input keys pressed)

#### Known issues
- ship models and weapon shells are not scalable, basing by screen size
- all speeds and not scalable by screen resolution
 
## [0.2-alpha] - 2023-09-16
  
Here we would have the update steps for 0.2-alpha for people to follow. This update is dedicated for game classes changes.  
The development was being made in OOP_Improvements branch.
 
### Added
- 'Game session' entity, which keeps data and settings of active session
- 'Score' attribute to count player's activity such as destroying enemies.
- Each WeaponShell (including Laser) now contain the information of the ship, who created it (parent)

 
### Changed
  
#### Tech
    
- Game elements are not belonged to each other. Instead, they become part of game session entity.
- Globally changed the behaviour of WeaponShell (incl. Laser). Now, they always make a damage (previously - destroyed ship if it was Enemy). Each ship now have health.
- Added Ship method to check of ship is still alive. If player makes a damage, and Enemy Ship's health drop to zero or below - Ship is removed from game session and disappears from screen
- Reduced code in inherited classes. Class Laser containt only 'pass' keyword inside. Reduced 'init' blocks in each class
- Methods 'draw' were removed elsewhere, except SpaceObject.
- Method 'shoot' was unified in parent Ship class and removed from EnemyShip and Player class.
- Method 'move' was removed in all classes except Base 
- Ship's attribute "ship_img" changed to "img" to correspond base SpaseObject class
- Since the lasers are no more part of ships, function 'move lasers" was moved to Main module. Cooldown was also moved there.
- 'Move' method in WeaponShell class was extended and added deep analysis parent/target to make sure that player's shells can attach only enemies, while enemies' ones - only player ship and no each other. (Required change since Lasers were moved to 'global scope' of game tech design - to 'game session').



#### Gameplay

- Reduced the cooldown of shooting for Player from 1/2 to 1/3 of second   

### Fixed
 
- Lasers are not disappearing when enemy ship is destroyed

### Known issues
  
- Function 'collide' is out of classes
  
## [0.1-alpha] - 2023-09-08
 
### Added
- Base game functionality based on Pygame module:  
You have to protect your own base (lives counter) and not allow any enemy to reach it. Your lifebar decreases if you get shot of collide with enemy.
- Constants module. Keeps color map and list of resolutions
- Base pause function (with limitations)
   
### Changed
- Game is being adapted for widescreen resolution
- Ship models were changed for hi-res models made by MidJourney
- Background and menu background images replaced for hi-res splash screens
- Classes are moved to dedicated module Data.

data.py
import os

import pygame

from const import COLORS

# Load images
RED_SPACE_SHIP = pygame.image.load(os.path.join('assets', 'enemy3_small.png'))
GREEN_SPACE_SHIP = pygame.image.load(os.path.join('assets', 'enemy2_small.png'))
BLUE_SPACE_SHIP = pygame.image.load(os.path.join('assets', 'enemy1_small.png'))

# Player player
YELLOW_SPACE_SHIP = pygame.image.load(os.path.join('assets', 'main_ship3_small.png'))

# Lasers
RED_LASER = pygame.image.load(os.path.join('assets', 'pixel_laser_red.png'))
GREEN_LASER = pygame.image.load(os.path.join('assets', 'pixel_laser_green.png'))
BLUE_LASER = pygame.image.load(os.path.join('assets', 'pixel_laser_blue.png'))
YELLOW_LASER = pygame.image.load(os.path.join('assets', 'pixel_laser_yellow.png'))

# Static objects
MEDKIT = pygame.image.load(os.path.join('assets', 'medkit_small.png'))


# class GameData:
#     '''Storage of game data'''

class SpaceObject: # need to create base class
    '''Base object that can appear on game screen'''
    def __init__(self, pos_x, pos_y, img):
        self.x = pos_x
        self.y = pos_y
        self.img = img
        self.mask = pygame.mask.from_surface(self.img)

    def draw(self, window):
        window.blit(self.img, (self.x, self.y))

    def move(self, vert_vel):
        self.y += vert_vel

    def collide(self, obj):
        offset_x = obj.x - self.x
        offset_y = obj.y - self.y
        return self.mask.overlap(obj.mask, (offset_x, offset_y)) != None


class WeaponShell(SpaceObject):
    '''Base shooting object, which can make a damage'''
    def __init__(self, pos_x, pos_y, img, parent, session_data, damage=10):
        super().__init__(pos_x, pos_y, img)
        self.damage: int = damage
        self.parent: Ship = parent
        self.session_data = session_data

    def off_screen(self, height):
        return not (self.y <= height and self.y >= 0) # ?

    def move(self, vert_vel, target=None, damage=10):
        super().move(vert_vel)
        exists = True
        if self.off_screen(self.session_data['screen_height']):
            exists = False
        elif isinstance(target, Player):
            # if self.collision(target):
            if self.collide(target):
                target.health -= damage
                exists = False
        elif isinstance(target, list):
            for obj in target:
                # if self.collision(obj):
                if self.collide(obj):
                    obj.health -= damage
                    exists = False
        return exists


class Laser(WeaponShell):
    '''Base attack shot, moved directly down'''
    pass


class StaticObject(SpaceObject):
    pass


class Medkit(StaticObject):
    def __init__(self, pos_x, pos_y):
        img = MEDKIT
        super().__init__(pos_x, pos_y, img)


class Ship(SpaceObject):
    COOLDOWN = 30 # half a second if FPS = 60

    def __init__(self, pos_x, pos_y, session_data, img=None, laser_img=None, health=100):
        super().__init__(pos_x, pos_y, img)
        self.health = health
        self.image = img
        self.laser_img = laser_img
        self.cool_down_counter = 0
        self.session_data = session_data

    def cooldown(self):
        if self.cool_down_counter >= self.COOLDOWN:
            self.cool_down_counter = 0
        elif self.cool_down_counter > 0:
            self.cool_down_counter += 1

    def shoot(self): # cooldown is located in main
        if self.cool_down_counter == 0:
            laser = Laser(self.x - ((self.laser_img.get_width() - self.img.get_width())/2), self.y, self.laser_img, self, self.session_data)
            self.session_data['weapon_shells'].append(laser)
            self.cool_down_counter = 1

    def is_dead(self):
        return self.health <= 0

    def get_width(self):
        return self.img.get_width()

    def get_height(self):
        return self.img.get_height()


class Player(Ship):
    COOLDOWN = 20 # 1/3 a second if fps = 60

    def __init__(self, pos_x, pos_y, session_data, player_choice=None, health=100):
        img = YELLOW_SPACE_SHIP
        laser_img = YELLOW_LASER
        super().__init__(pos_x, pos_y, session_data, img, laser_img, health)
        self.max_health = health

    def draw(self, window):
        super().draw(window)
        self.healthbar(window) # could improve with mixin

    def healthbar(self, window): # redraw to improve graphics, use resize
        pygame.draw.rect(window, COLORS['red'],
                         (self.x, self.y + self.img.get_height() + 10,
                         self.img.get_width(), 10))
        pygame.draw.rect(window, COLORS['green'],
                         (self.x, self.y + self.img.get_height() + 10,
                         self.img.get_width() * (self.health / self.max_health), 10))


class Enemy(Ship):
    COLOR_MAP = { # could be types later
        'red': (RED_SPACE_SHIP, RED_LASER),
        'green': (GREEN_SPACE_SHIP, GREEN_LASER),
        'blue': (BLUE_SPACE_SHIP, BLUE_LASER)
    }

    def __init__(self, pos_x, pos_y, session_data, color, health=10):
        img, laser_image = self.COLOR_MAP[color]
        super().__init__(pos_x, pos_y, session_data, img, laser_image, health)




main.py
import pygame

from const import COLORS
from data import Player
from helper import (game_init, game_session_init, paused, escape_game,
                    get_horizontal_center_position, get_middle_position,
                    generate_enemies, enemies_movement, weapon_shell_movement,
                    generate_medkits, static_objects_movement)

def main(session_data):
    run = True

    main_font = pygame.font.SysFont('comicsans', game_data['height']//18)
    lost_font = pygame.font.SysFont('comicsans', game_data['height']//15)
    
    player = Player(game_data['width'] / 2,
                    game_data['height'] * 0.87, # magic
                    session_data)

    clock = pygame.time.Clock()

    lost = False
    lost_count = 0 # works for 'defeat' message stays on screen
    
    def redraw_window():
        game_data['WINDOW'].blit(game_data['BG'], (0, 0))
        # draw text
        lives_label = main_font.render(f'Lives: {session_data["lives"]}', 1, COLORS['white'])
        level_label = main_font.render(f'Level: {session_data["level"]}', 1, COLORS['white'])
        score_label = main_font.render(f'Score: {session_data["score"]}',
                                              1, COLORS['white'])

        game_data['WINDOW'].blit(lives_label, (10, 10))
        game_data['WINDOW'].blit(level_label, (game_data['width'] - level_label.get_width() - 10, 10))
        game_data['WINDOW'].blit(score_label, 
                    (get_horizontal_center_position(score_label), 10))
        # debug_label = main_font.render('DEBUG========0.7===================++++++++++++', 1, COLORS['white'])
        # game_data['WINDOW'].blit(debug_label, (0, game_data['height'] * 0.7))

        
        for enemy in session_data['enemies']:
            enemy.draw(game_data['WINDOW'])

        for shell in session_data['weapon_shells']:
            shell.draw(game_data['WINDOW'])

        for obj in session_data['static_objects']:
            obj.draw(game_data['WINDOW'])

        player.draw(game_data['WINDOW'])

        if lost:
            lost_label = lost_font.render('You have lost!', 1, COLORS['white'])
            game_data['WINDOW'].blit(lost_label, get_middle_position(lost_label))

        pygame.display.update()

    while run:
        clock.tick(game_data['FPS'])
        redraw_window()

        if session_data['lives'] <= 0 or player.is_dead():
            lost = True
            lost_count += 1
        
        if lost:
            if lost_count > game_data['FPS'] * 3: # 60 fps * 3 means 3 sec
                run = False
            else:
                continue

        # Enemies generation
        if len(session_data['enemies']) == 0:
            generate_enemies(session_data)

        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                quit()
            elif event.type == pygame.KEYDOWN:
                if event.key == pygame.K_p:
                    paused()
                elif event.key == pygame.K_ESCAPE:
                    escape_game()


        keys = pygame.key.get_pressed() # possible to add info on I and help on H
        if (keys[pygame.K_a] or keys[pygame.K_LEFT]) and player.x - session_data['player_vel'] > 0: #left
            player.x -= session_data['player_vel']
        if (keys[pygame.K_d] or keys[pygame.K_RIGHT]) and player.x + session_data['player_vel'] + player.get_width() < game_data['width']: # right, size player
            player.x += session_data['player_vel']
        if (keys[pygame.K_w] or keys[pygame.K_UP]) and player.y - session_data['player_vel'] > 0: #up
            player.y -= session_data['player_vel']
        if (keys[pygame.K_s] or keys[pygame.K_DOWN]) and player.y + session_data['player_vel'] + player.get_height() + 18 < game_data['height']: # down, size player, 18 to show healthbar
            player.y += session_data['player_vel']
        if keys[pygame.K_SPACE]:
            player.shoot()

        # Weapon shell movement
        weapon_shell_movement(session_data, player)

        # Player cooldown for shooting
        player.cooldown() # extracted from player.move_lasers

        # Enemy movement
        enemies_movement(session_data, player)

        # Generate medkits
        generate_medkits(session_data, player)

        # Static objects movements
        static_objects_movement(session_data, player)



def main_menu():
    global game_data
    title_font = pygame.font.SysFont('comicsans', int(game_data['height']/12.85))
    run = True
    while run:
        game_data['WINDOW'].blit(game_data['MENU_BG'], (0, 0))
        title_label = title_font.render('Press the mouse to begin...', 1, COLORS['white'])
        game_data['WINDOW'].blit(title_label, get_middle_position(title_label))

        pygame.display.update()
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                run = False
            if event.type == pygame.MOUSEBUTTONDOWN:
                session_data = game_session_init() # possibly to modify in future with difficulty level
                main(session_data)
            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_ESCAPE:
                    game_data['WINDOW'].blit(game_data['BG'], (0, 0))
                    pygame.display.update()
                    escape_game()
    pygame.quit()


if __name__ == '__main__':
    game_data = game_init()
    main_menu()


helper.py
import os
from typing import List #, Dict, Type

import pygame
import random
from screeninfo import get_monitors, Monitor

from const import COLORS, WS_RESOLUTIONS
from data import Enemy, Medkit, Player, StaticObject, WeaponShell


def get_screen_res():
    monitors: List[Monitor] = get_monitors()
    width, height = monitors[0].width, monitors[0].height
    display: int = 0
    final_resolution = None
    for i in range(len(monitors)):
        if monitors[i].is_primary == True:
            width = monitors[i].width
            height = monitors[i].height
            display = i
    baseheight = width / 16 * 9
    if height <= baseheight: # 16:9 monitor or less (4:3)
        # height is max
        for res in WS_RESOLUTIONS:
            if WS_RESOLUTIONS[res][1] <= height:
                final_resolution = res
                break
    else: # superwide
        # width is max
        for res in WS_RESOLUTIONS:
            if WS_RESOLUTIONS[res][0] <= width:
                final_resolution = res
                break
    if not final_resolution:
        final_resolution = '1280x720 (HD)'
    # print('DEBUG', final_resolution)
    return WS_RESOLUTIONS[final_resolution], display, final_resolution


def game_init():
    ((screen_width, screen_height), display, final_res) = get_screen_res()
    pygame.font.init() # maybe pygame.init?
    global game_data
    pygame.display.set_caption('Space Arcade')

    game_data = {
        'resolution': WS_RESOLUTIONS[final_res],
        # 'music': False,
        # 'sounds': False,
        'FPS': 60,
        'width': screen_width,
        'height': screen_height,
        'display': display,
        'WINDOW': pygame.display.set_mode((screen_width, screen_height)),
        'BG': pygame.transform.scale(pygame.image.load(os.path.join('assets', 'space_bg.webp')), (screen_width, screen_height)),
        'MENU_BG': pygame.transform.scale(pygame.image.load(os.path.join('assets', 'main_menu.webp')), (screen_width, screen_height))

    }
    return game_data

def game_session_init():
    session_data = {
        'score': 0,
        'enemies': [],
        'weapon_shells': [],
        'static_objects': [],
        'level': 0,
        'lives': 5,
        'wave_length': 5,
        'enemy_vel': 1,
        'player_vel': game_data['width'] // 320, # 5 in 1600*900
        'laser_vel': 5,
        'laser_damage': 10,
        'static_vel': 2,
        'screen_height': game_data['height']
    }

    return session_data

def get_horizontal_center_position(obj):
    global game_data
    return game_data['width'] / 2 - obj.get_width() / 2

def get_vertical_center_position(obj):
    return game_data['height'] / 2 - obj.get_height() / 2

def get_middle_position(obj):
    return (get_horizontal_center_position(obj),
            get_vertical_center_position(obj))

def paused():
    global game_data
    pause: bool = True
    pause_font = pygame.font.SysFont('comicsans', 70)

    while pause:
        pause_label = pause_font.render('Pause. Press any key to continue', 1, COLORS['violet'])
        game_data['WINDOW'].blit(pause_label, get_middle_position(pause_label))
        pygame.display.update()
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                quit()
            elif event.type == pygame.KEYDOWN:
                pause = False

def escape_game():
    global game_data
    pause: bool = True
    escape_font = pygame.font.SysFont('comicsans', 70)

    while pause:
        pause_label_first = escape_font.render('Are you going to leave game?', 1, COLORS['violet'])
        pause_label_second = escape_font.render('Press Enter to end the game or ESC to return', 1, COLORS['violet'])
        game_data['WINDOW'].blit(pause_label_first, (get_horizontal_center_position(pause_label_first), game_data['height']/2 - pause_label_first.get_height())) # was HEIGHT
        game_data['WINDOW'].blit(pause_label_second, (get_horizontal_center_position(pause_label_second), game_data['height']/2 + pause_label_first.get_height())) # was WIDTH
        pygame.display.update()
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                quit()
            elif event.type == pygame.KEYDOWN:
                if event.key == pygame.K_ESCAPE:
                    pause = False
                elif event.key == pygame.K_RETURN or event.key == pygame.K_KP_ENTER:
                    pygame.quit()
                    quit()


def generate_enemies(session_data):
    session_data['level'] += 1
    session_data['wave_length'] += 5 # add 5 more enemies each level
    for _ in range(session_data['wave_length']):
        enemy = Enemy(random.randrange(50, game_data['width'] - 100),
                        random.randrange(-1500, -100), # -1500*level/5
                        session_data,
                        random.choice(['red', 'blue', 'green']))
        session_data['enemies'].append(enemy)


def enemies_movement(session_data, player: Player): # possible to remove player if move collide inside Space object class
    enemy: Enemy
    for enemy in session_data['enemies'][:]:
        if enemy.is_dead():
            session_data['enemies'].remove(enemy)
            session_data['score'] += 1
            continue
        enemy.move(session_data['enemy_vel'])
        enemy.cooldown() # refresh cooldown to shoot

        if (enemy.y <= game_data['height'] * 0.7 and
            random.randrange(0, 2*game_data['FPS']) == 1): # each ~ 2 sec
            enemy.shoot()

        if enemy.collide(player):
            player.health -= 10
            session_data['score'] += 1
            session_data['enemies'].remove(enemy)
        elif enemy.y + enemy.get_height() > game_data['height']:
            session_data['lives'] -= 1
            session_data['enemies'].remove(enemy)


def weapon_shell_movement(session_data, player: Player) -> None:
    shell: WeaponShell
    for shell in session_data['weapon_shells'][:]:
        if shell.parent == player:
            target = session_data['enemies']
            speed = -session_data['laser_vel']
        else:
            target = player
            speed = session_data['laser_vel']
        shell_exists = shell.move(speed, target, session_data['laser_damage'])
        if not shell_exists:
            session_data['weapon_shells'].remove(shell)

def generate_medkits(session_data, player: Player):
    probability = 30 # each 30 sec
    if player.health <= 20:
        probability = 10
    elif player.health <= 70:
        probability = 20
    if random.randrange(0, probability*game_data['FPS']) == 1:
        medkit = Medkit(random.randrange(50, game_data['width'] - 100),
                        random.randrange(-1500, -100))
        session_data['static_objects'].append(medkit)


def static_objects_movement(session_data, player: Player):
    obj: StaticObject
    for obj in session_data['static_objects'][:]:
        obj_h_vel = session_data['static_vel']
        if obj.y > game_data['height'] * 0.7:
            obj_h_vel = random.randrange(-session_data['static_vel'],
                                       0,3 * session_data['static_vel'])
        obj.move(obj_h_vel)
        if obj.collide(player):
            player.health += 50
            if player.health >= player.max_health:
                player.health = player.max_health
            session_data['static_objects'].remove(obj)
